File: plugins/geolocation-plugin/ios/Plugin/GeolocationCoordinatesPlugin.swift
import Foundation
import Capacitor
import CoreLocation

/**
 * Geolocation plugin
 */

class GetLocationHandler: NSObject, CLLocationManagerDelegate {
  var locationManager = CLLocationManager()
  var call: CAPPluginCall

  init(call: CAPPluginCall) {
    self.call = call
    
    super.init()
    
    self.locationManager.delegate = self
    self.locationManager.requestWhenInUseAuthorization()
    if call.getBool("enableHighAccuracy", false) {
        self.locationManager.desiredAccuracy = kCLLocationAccuracyBestForNavigation
    } else {
      self.locationManager.desiredAccuracy = kCLLocationAccuracyThreeKilometers
    }
    
    self.locationManager.startUpdatingLocation()
   
  }

  public func stopUpdating() {
    self.locationManager.stopUpdatingLocation()
  }
  
  func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
    call.error(error.localizedDescription, error, [
      "message": error.localizedDescription
    ])
  }
  
  func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
    if let location = locations.first  {
      let result = makePosition(location)
      call.success(result)
    } else {
      call.success()
    }
  }
  
  func makePosition(_ location: CLLocation) -> JSObject {
    var geoLocation = JSObject()
    geoLocation["lat"] = location.coordinate.latitude
    geoLocation["long"] = location.coordinate.longitude
    return geoLocation
  }
}

@objc(GeolocationCoordinatesPlugin)
public class GeolocationCoordinatesPlugin: CAPPlugin {
    private let implementation = GeolocationCoordinates()
    
    var locationHandler: GetLocationHandler?
    @objc func getCoordinates(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
         self.locationHandler = GetLocationHandler(call: call)
       }
     }
    
    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }
}

File: plugins/geolocation-plugin/android/src/main/java/com/test/geolocation/GeolocationCoordinatesPlugin.java
package com.test.geolocation;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "GeolocationCoordinates")
public class GeolocationCoordinatesPlugin extends Plugin {

    private GeolocationCoordinates implementation = new GeolocationCoordinates();
    private GpsTracker gpsTracker;

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");

        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }

    @PluginMethod
    public void getCoordinates(PluginCall call) {
        gpsTracker = new GpsTracker(this.getContext());
        if(gpsTracker.canGetLocation()){
            JSObject obj = new JSObject();
            double latitude = gpsTracker.getLatitude(); 
            double longitude = gpsTracker.getLongitude();
            obj.put("lat", latitude);
            obj.put("long", longitude);
            call.resolve(obj);
        }else{
            gpsTracker.showSettingsAlert();
        }
    }
}
